Feature/10 - UserSettings

valerachuk opened the pull request:
![image](https://user-images.githubusercontent.com/38440869/82507300-8b768500-9b0a-11ea-9bc0-4c466067bf66.png)
Can't override parent's !important padding

valerachuk committed:
add user settings

simple scenario works
add mmt-dense to mmt-text-input
add settings mock data

valerachuk committed:
Merge branch 'dev' into feature/10

valerachuk committed:
validation + localization :book:

valerachuk committed:
Merge branch 'dev' into feature/10

valerachuk committed:
move settings to user panel (/my)

horba commented on front/package.json:
```diff
@@ -12,9 +12,9 @@
-    "vuex": "^3.1.3"
+    "vue-router": "^3.2.0",
+    "vuetify": "^2.2.28",
+    "vuex": "^3.4.0"
```
ну вообще как-то нехорошо вышло

horba commented on front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html:
```diff
@@ -0,0 +1,3 @@
+<v-btn class="mmt-continue-button" rounded depressed large v-bind="attrs" v-on="$listeners">
+  <slot></slot>
+</v-btn>
```
не думаю, что стоит выносить кнопку в отдельный компонент

demosglok replied:
why are we binding to attrs here?

horba replied:
концепция прозрачного компонента враппера

horba replied:
правда, здесь я не вижу в ней смысла

valerachuk replied:
Не нужно будет по несколько раз прописывать одни и те же стили и атрибуты форматирования. Такая кнопка есть в форме логина и в форме настроек. Тогда я верну ее обратно

horba commented on front/src/config/config.dev.json:
```diff
@@ -1,3 +1,4 @@
 {
-  "serverPath": "https://localhost:5001/"
+  "serverPath": "https://localhost:5001/",
+  "emailRegex": "^\\w+([\\.-]?\\w+)*@\\w+([\\.-]?\\w+)*(\\.\\w{2,3})+$"
```
why here? actually this file is not for those purposes

valerachuk replied:
Should I create a config file for validation rules ?

horba replied:
it is not config. if you want you can create file with constants

horba commented on front/src/locales/translations/ua.json:
```diff
@@ -39,5 +39,53 @@
+        "Ukrainian": "Українська",
+        "German": "Німецька"
+      }
+    },
```
we have had these translations already

horba commented on front/src/locales/translations/ua.json:
```diff
@@ -39,5 +39,53 @@
+      }
+    },
+    "country": {
+      "Label": "Країна",
```
it is way to promote copypast in future. Is is vocabulary. Keys can be logical, but not ui-based. No label, options key

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+        .then(() => console.log('Settings updated'))
+        .catch(error => {
+          this.emailAlreadyTaken = error.response.data === 'Email already taken';
+        });
```
action in store

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+    return {
+      genders: [],
+      languages: [],
+      countries: [],
```
are you sure that it is not in store already?

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+    fillLoadedFields (data) {
+      this.genders = this.convertToSelectCompatible(data.genders, 'gender');
+      this.languages = this.convertToSelectCompatible(data.languages, 'language');
+      this.countries = this.convertToSelectCompatible(data.countries, 'country');
```
it is vocabularies that are used not only in your component, and each time front will request the same data

horba replied:
also I can't understand why you need converters

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+          text: this.$t(`userSettings.${fieldType}.options.${id}`)
+        }));
+    }
+  },
```
flexibility vs maintainable

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+        this.fillLoadedFields(data);
+        this.isLoaded = true;
+      });
+  }
```
action in store

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+      languages: [],
+      countries: [],
+      lastName: '',
+      firstName: '',
```
don't we need these data anywhere in application?

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+      firstName: '',
+      email: '',
+      gender: '',
+      language: '',
```
is these one connected whith localization? Should be

valerachuk replied:
These fields are containing ids, translation generates here: convertToSelectCompatible()

horba replied:
yes, as I said it may make feeling of flexibility, but from other side it makes code more difficult for reading and support

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,73 @@
+      countries: [],
+      lastName: '',
+      firstName: '',
+      email: '',
```
actually, email data is usually necessary not only for user settings

horba commented on server/WebAPI/DTO/UserSettings/UserSettingsResponse.cs:
```diff
@@ -0,0 +1,35 @@
+
+namespace WebAPI.DTO.UserSettings
+{
+    public class UserSettingsResponse
```
UserSettingsDTO, see my comment for Irina

horba replied:
http://joxi.ru/Vm6ok6Vsj9eYX2

horba reviewed (changes requested):
component - is about user interaction, logic part and data management are in store

demosglok commented on front/src/components/controls/mmt-select/mmt-select.component.js:
```diff
@@ -0,0 +1,9 @@
+export default {
+  inheritAttrs: false,
+  computed: {
+    attrs () {
```
do we really need it?

demosglok commented on front/src/components/controls/mmt-submit-button/mmt-submit-button.component.js:
```diff
@@ -0,0 +1,8 @@
+export default {
+  computed: {
+    attrs () {
+      const { rounded, depressed, large, ...other } = this.$attrs;
```
why do we need it?

demosglok commented on front/src/components/controls/mmt-text-input/mmt-text-input.component.css:
```diff
@@ -1,8 +1,16 @@
 
+label,
+.mmt-text-input >>> .v-messages__message {
+  font-size: 12px;
```
will css-variable be more preferable here?

demosglok commented on front/src/views/user/user-settings/user-settings.view.html:
```diff
@@ -0,0 +1,22 @@
+    <h5 class="form-header">{{$t('userSettings.title')}}</h5>
+  </v-card-title>
+  <v-card-text>
+    <v-form class="settings-form" ref="form" v-model="formValidity">
```
do we really need "ref" here?

demosglok commented on server/WebAPI/Controllers/UserSettingsController.cs:
```diff
@@ -0,0 +1,67 @@
+        {
+            if (!Int32.TryParse(HttpContext.User.Claims.FirstOrDefault(x => x.Type == ClaimTypes.Name)?.Value, out var id))
+            {
+                return Forbid();
```
Forbidden?

valerachuk replied:
Yes, user can't get settings if id is not represented in JWT

horba replied:
Дмитрий говорит, что называться должно Forbidden, а не Forbid

valerachuk replied:
В ControllerBase нет метода Forbidden, есть Forbid

demosglok commented on server/WebAPI/Controllers/UserSettingsController.cs:
```diff
@@ -0,0 +1,67 @@
+
+            if (userWithThisEmail != null && userWithThisEmail.Id != id)
+            {
+                return BadRequest("Email already taken");
```
Email IS taken:)

valerachuk committed:
move userSerrings to store

fix userSettings locale
move emailRegex to constants.json
revert npm update
rename UserRequest to DTO

valerachuk committed:
Merge branch 'dev' into feature/10

valerachuk committed:
Fix user settings (lang + country)

move language from enum to entity
connect country setting with country entity

valerachuk commented:
![image](https://user-images.githubusercontent.com/38440869/82770128-04921700-9e40-11ea-8e4e-da6f9e32e0b5.png)
Some warnings :(

horba commented on front/src/main.js:
```diff
@@ -6,6 +6,7 @@ import vuetify from './plugins/vuetify';
 import i18n from './plugins/i18n';
 import axios from 'axios';
 import { Default, Authorize } from '@layout';
+import { fetchServer } from '@/services/fetchServer.js';
```
add path `@services`

horba commented on front/src/main.js:
```diff
@@ -35,6 +36,11 @@ new Vue({
     });
+
+    if (this.$store.getters.isLoggedIn) {
+      // get userSettings data from server
```
these comments usually means that your code is understandable

horba replied:
1) it is really strange name fetchServer if you fetch actually user settings
2) i don't agree that only one method is enough and has sense to place into separate service

horba commented on front/src/services/fetchServer.js:
```diff
@@ -0,0 +1,5 @@
+import store from '@store';
+
+export const fetchServer = () => {
+  store.dispatch('userSettings/loadUserSettings');
```
why did you do it? it is really confused me

valerachuk replied:
In future, here we can add actions with the same functionality

horba commented on front/src/views/sign-in/sign-in.view.js:
```diff
@@ -21,6 +23,7 @@ export default {
       })
         .then(() => {
+          fetchServer();
           this.$router.push('/');
```
routing push better to place into component. it is interaction.

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,149 @@
+    },
+    ...mapGetters('userSettings', [ 'userSettings' ])
+  },
+  watch: {
```
why did you need watch?

valerachuk replied:
To fill form fields when data becomes available

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,149 @@
+    }
+  },
+  methods: {
+    ...mapActions('userSettings', [ 'updateUserSettings' ]),
```
why need you map store methods?

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,149 @@
+        }
+      )
+        .catch(error => {
+          this.emailIsAlreadyTaken = error.response.data === 'Email is already taken';
```
it should be checked not via text. you can use error numbers or others

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,149 @@
+      this.email = data.email;
+      this.firstName = data.firstName;
+      this.lastName = data.lastName;
+      this.isLoaded = true;
```
why did you need isLoaded?

valerachuk replied:
To disable form, while form data is loading

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,149 @@
+    }
+  },
+  created () {
+    // if settings are available in store - fill fields
```
you have computed properties from store.

horba reviewed (changes requested):
Might, you need to read about store and computed properties

horba commented on front/src/views/user/user-settings/user-settings.view.js:
```diff
@@ -0,0 +1,149 @@
+  created () {
+    // if settings are available in store - fill fields
+    if (this.userSettings != null) {
+      this.fillLoadedFields(this.userSettings);
```
```
data() {
editedUserSettings: {}
}
```
=======
`this.editedUserSettings = Object.assign({}, this.userSettings)`

=============

`<form v-if="editedUserSettings"/>`

https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

valerachuk replied:
> this.editedUserSettings = Object.assign({}, this.userSettings)

If form creates before its data loads, editedUserSettings will be empty. So its better to place this line inside watch

valerachuk replied:
Object is always true...
I can use isLoaded or editedUserSettings.email to check form loading status

valerachuk replied:
Used computed property: 

```javascript
isLoaded () {
      return !!Object.keys(this.editedUserSettings).length;
    }
```

horba commented on front/src/store/modules/userSettings.js:
```diff
@@ -0,0 +1,31 @@
+  },
+  getters: {
+    userSettings (state) {
+      return state.userSettings;
```
do we really need this getter?

valerachuk replied:
No, I forget to delete it

horba commented on front/src/views/sign-in/sign-in.view.js:
```diff
@@ -21,6 +23,7 @@ export default {
         password: this.password
       })
         .then(() => {
+          fetchServer();
```
why here?
1) you can just call dispatch action and it will be more clear like store.dispatch('userSettings/loadUserSettings'); 
2) why here we need to init user settings?it is last place where i will looking for init of user data. i expect that we call it in store on successfull authorize or from base authorized layout, but not here

valerachuk replied:
1. already here
2. ok, I will place it inside layout

valerachuk committed:
fix explicit issues

delete useless service
delete mapActions from user-settings.vue.js
add JWT id getters to UserSettingsController from extention

valerachuk committed:
Merge branch 'dev' into feature/10

valerachuk committed:
fix sign-up text fields after merge

add mmt-dense

valerachuk committed:
user settings code refactor

fix login form color
move user settings loader dispatcher to authorize layout
